### PR TITLE
[data/db] SDP-262: Retry invitation SMS

### PR DESCRIPTION
**Pull request by ceciliaromao**
### What

Added a timestamp in the receiver_wallets table called invitation_sent_at.

### Why

We can send invitations based on rows that have that field empty, and we can trigger retries by clearing that field.
This unblocks the front-end ticket. 

### Known limitations

@CaioTeixeira95 will address the new column in his next PRs.

#### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR does not mix refactoring changes with feature changes (split into two PRs otherwise).
* [x] This PR's title starts with the name of the package, area, or subject affected by the change.

#### Thoroughness

* [x] This PR adds tests for the new functionality or fixes.
* [ ] This PR contains the link to the Jira ticket it addresses.

#### Configs and Secrets

* [x] No new CONFIG variables are required -OR- the new required ones were added to the helmchart's [`values.yaml`] file.
* [x] No new CONFIG variables are required -OR- the new required ones were added to the deployments ([`pr-preview`], [`dev`], [`demo`], `prd`).
* [x] No new SECRETS variables are required -OR- the new required ones were mentioned in the helmchart's [`values.yaml`] file.
* [x] No new SECRETS variables are required -OR- the new required ones were added to the deployments ([`pr-preview secrets`], [`dev secrets`], [`demo secrets`], `prd secrets`).

#### Release

* [x] This is not a breaking change.
* [x] **This is ready for production.**. If your PR is not ready for production, please consider opening additional complementary PRs using this one as the base. Only merge this into `develop` or `main` after it's ready for production!

#### Deployment

* [ ] Does the deployment work after merging?

[`values.yaml`]: ../helmchart/sdp/values.yaml
[`pr-preview`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/stellar-disbursement-platform/backend-helm-values
[`dev`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/backend-helm-values
[`demo`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-backend-helm-values
[`pr-preview secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/externalsecrets-common-previews.yaml#L241-L346
[`dev secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/stellar-disbursement-platform-externalsecrets.yaml
[`demo secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-sdp-externalsecrets.yaml

### CODE REVIEW

File: internal/data/receivers_wallet.go
Comment (marcelosalloum) [inline internal/data/receivers_wallet.go:502]: The order of the method signature we try to use throughout the project starts with the dependencies and then the input params, like in `context, sqlExec, receiverWalletId`. In other words, dependencies come first and input parameters later.

```suggestion
func (rw *ReceiverWalletModel) RetryInvitationSMS(ctx context.Context, sqlExec db.SQLExecuter, receiverWalletId string) (*ReceiverWallet, error) {
```

File: internal/db/migrations/2023-09-29.0-alter-receiver_wallets-add-invitation_sent_at-column.sql
Comment (marcelosalloum) [inline internal/db/migrations/2023-09-29.0-alter-receiver_wallets-add-invitation_sent_at-column.sql:5]: 🚩Your migration should also populate that column with a non null value for receiver wallets who were already invited. Otherwise, you will end up triggering new SMS for existing users who already received an invitation.

Please, keep in mind this product is already running in production and any changes can affect real users.

File: internal/serve/httphandler/receiver_wallets_handler.go
Comment (marcelosalloum) [inline internal/serve/httphandler/receiver_wallets_handler.go:33]: Calls to `httperror.InternalError` already trigger a log-and-report error call, so this log here is redundant.

```suggestion
		httperror.InternalError(ctx, "", err, nil).Render(rw)
```

Comment (marcelosalloum) [reply internal/serve/httphandler/receiver_wallets_handler.go:33]: Here are the code blocks that setup that behavior:

https://github.com/stellar/stellar-disbursement-platform-backend/blob/0aabb19139bd9c0699a53e6a6b6319ec624f9383/internal/serve/httperror/httperror.go#L119

https://github.com/stellar/stellar-disbursement-platform-backend/blob/0aabb19139bd9c0699a53e6a6b6319ec624f9383/internal/serve/serve.go#L94-L95

https://github.com/stellar/stellar-disbursement-platform-backend/blob/0aabb19139bd9c0699a53e6a6b6319ec624f9383/internal/serve/httperror/httperror.go#L25-L38

**Review by marcelosalloum — COMMENTED** [review]
Thank you for working on that.

This seems to unblock the frontend ticket but when will this new column be used? Will @CaioTeixeira95 address that in his next PRs, or are you planning to do that after this one is merged? This kind of info should be present in the pull request description, so code reviewers know what to review.

File: internal/serve/httphandler/receiver_wallets_handler.go
Comment (CaioTeixeira95) [inline internal/serve/httphandler/receiver_wallets_handler.go:18]: There's no need to receive the `DBConnectionPool` here, you can get it from the `data.Models`.

```suggestion
type ReceiverWalletsHandler struct {
	Models           *data.Models
}
```

File: internal/serve/httphandler/receiver_wallets_handler.go
Comment (CaioTeixeira95) [inline internal/serve/httphandler/receiver_wallets_handler.go:24]:
```suggestion
	receiverWallet, err := h.Models.ReceiverWallet.RetryInvitationSMS(ctx, h.Models.DBConnectionPool, receiverWalletID)
```

File: internal/serve/httphandler/receiver_wallets_handler.go
Comment (CaioTeixeira95) [inline internal/serve/httphandler/receiver_wallets_handler.go:30]: This error message seems incorrect in this context. Can you please update it?

File: internal/serve/httphandler/receiver_wallets_handler_test.go
Comment (CaioTeixeira95) [inline internal/serve/httphandler/receiver_wallets_handler_test.go:41]:
```suggestion
		req, err := http.NewRequestWithContext(ctx, http.MethodPatch, route, nil)
```

File: internal/serve/httphandler/receiver_wallets_handler_test.go
Comment (CaioTeixeira95) [inline internal/serve/httphandler/receiver_wallets_handler_test.go:48]: Could you please assert the response body as well?

File: internal/serve/httphandler/receiver_wallets_handler_test.go
Comment (CaioTeixeira95) [inline internal/serve/httphandler/receiver_wallets_handler_test.go:57]:
```suggestion
		req, err := http.NewRequestWithContext(ctx, http.MethodPatch, route, nil)
```

File: internal/serve/httphandler/receiver_wallets_handler_test.go
Comment (CaioTeixeira95) [inline internal/serve/httphandler/receiver_wallets_handler_test.go:64]: Here too. Could you assert the response body?

**Review by CaioTeixeira95 — COMMENTED** [review]
Great job! I've left some suggestions.

Also, I want to bring here the necessity to retry the invitation SMS when a new `payment` is created for receivers who never got registered for the wallet of the disbursement.

cc: @marcelosalloum

**Comment by CaioTeixeira95** [comment]
> This seems to unblock the frontend ticket but when will this new column be used? Will @CaioTeixeira95 address that in his next PRs, or are you planning to do that after this one is merged? 

That's correct, I'm going to use it in my next PR. @ceciliaromao could you please add a comment in the `Known Limitations` section about this?

File: internal/db/migrations/2023-09-29.0-alter-receiver_wallets-add-invitation_sent_at-column.sql
Comment (CaioTeixeira95) [inline internal/db/migrations/2023-09-29.0-alter-receiver_wallets-add-invitation_sent_at-column.sql:1]: Can you update the filename to the latest date before the merge?

File: internal/db/migrations/2023-09-29.0-alter-receiver_wallets-add-invitation_sent_at-column.sql
Comment (CaioTeixeira95) [inline internal/db/migrations/2023-09-29.0-alter-receiver_wallets-add-invitation_sent_at-column.sql:16]: It's not a big deal, but it's a good practice to have an empty line at the end of files.

**Review by CaioTeixeira95 — APPROVED** [review]

**Pull request merged**
